apiserver/crossmodel/crossmodel.go

```diff
@@ -60,44 +61,63 @@ func NewAPI(
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	offeredApplications := state.NewOfferedApplications(st)
-	backend := &servicesBackend{
-		offeredApplications,
-		applicationOffers,
-	}
-	return createAPI(backend, getStateAccess(st), resources, authorizer)
+	return createAPI(applicationOffers, getStateAccess(st), authorizer)
 }
 
-// Offer makes service endpoints available for consumption.
+// Offer makes application endpoints available for consumption at a specified URL.
 func (api *API) Offer(all params.ApplicationOffersParams) (params.ErrorResults, error) {
 	result := make([]params.ErrorResult, len(all.Offers))
+	var offers params.AddApplicationOffers
+	badOfferIndexes := set.NewInts()
 	for i, one := range all.Offers {
-		offer, applicationOfferParams, err := api.makeOfferedApplicationParams(one)
+		applicationOfferParams, err := api.makeOfferedApplicationParams(one)
 		if err != nil {
 			result[i].Error = common.ServerError(err)
+			badOfferIndexes.Add(i)
 			continue
 		}
-		if err := api.backend.AddOffer(offer, params.AddApplicationOffer{
+		offers.Offers = append(offers.Offers, params.AddApplicationOffer{
 			ApplicationOffer: applicationOfferParams,
 			UserTags:         one.AllowedUserTags,
-		}); err != nil {
-			result[i].Error = common.ServerError(err)
+		})
+	}
+	errResults, err := api.applicationDirectory.AddOffers(offers)
+	if err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+
+	// Merge the results with any errors encountered earlier.
```

Comment:
The merge process is a bit hard to follow. What about establishing a map of index-in-offers.Offers to index-in-result in the first loop? Then `badOfferIndexes` wouldn't be required and the results from `AddOffers` could be slotted straight into `errResults.Results`.